### PR TITLE
Cass client pool refactor

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
The `CassandraClientPool` needs refactoring in order to be functional with the Cassandra server deployed in k8s infra.
This PR lays the groundwork of this refactor. 

**Implementation Description (bullets)**:
We introduce a new model `CassandraServer` which contains the actual cassandraHost Ip and the proxy Ip which we will use to talk the cassandraHost. 
In this PR all of the comparisons (for blacklisting, absentHostTracker etc) are done on the cassandraHost returned by the `describe_ring` endpoint. The actual call to Cassandra are made using a proxy.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Relying on existing tests for sanity test. We are designing a sensible e2e test on our internal dev stack.

**Concerns (what feedback would you like?)**:
Does the design look sensible? 
Did I destroy an invariant?

**Where should we start reviewing?**:
CassandraServer, CassandraService

**Priority (whenever / two weeks / yesterday)**:
yesterday 🏃🏼‍♀️

### CODE REVIEW

**Comment by sudiksha27** [comment]
I do not really want to merge the PR as it is. I am raising it for the easy of reviewing. I would be cutting another branch from this which will incorporate changes to the way `CassandraServer` will be used.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 1]
```diff
@@ -68,21 +73,22 @@
 
     public CassandraClientPoolingContainer(
             MetricsManager metricsManager,
-            InetSocketAddress host,
+            CassandraServer cassandraServer,
             CassandraKeyValueServiceConfig config,
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java:77]: cassandraServer is only used for logging. You can see that the impl of `CassandraClientFactory` has not changed. We pass in the proxy field to `CassandraClientFactory` that creates a client.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java [hunk 2]
```diff
@@ -138,7 +141,7 @@ public String describeBlacklistedHosts() {
         return blacklist.entrySet().stream()
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java:145]: only called from trace log is we cannot find targeted host for key

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 3]
```diff
@@ -365,52 +363,55 @@ void runOneTimeStartupChecks() {
...
+        if (completelyUnresponsiveNodes.size() > 0) {
             errorBuilderForEntireCluster
                     .append("Performing routine startup checks,")
                     .append(" determined that the following hosts are unreachable for the following reasons: \n");
-            completelyUnresponsiveHosts.forEach((host, exception) -> errorBuilderForEntireCluster.append(String.format(
-                    "\tHost: %s was marked unreachable" + " via exception: %s%n",
-                    host.getHostString(), exception.toString())));
+            completelyUnresponsiveNodes.forEach(
+                    (cassandraServer, exception) -> errorBuilderForEntireCluster.append(String.format(
+                            "\tServer: %s was marked unreachable via proxy: %s, " + " with exception: %s%n",
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:397]: It is a one time start up check, even if too long we fail and do not start up so ....the worst has already happened.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 4]
```diff
@@ -505,17 +506,21 @@ private void removeFromBlacklistAfterResponse(InetSocketAddress host) {
     // think it does and they will not send writes to it, but it will respond to requests
     // acting like it does.
     private void sanityCheckRingConsistency() {
-        Multimap<Set<TokenRange>, InetSocketAddress> tokenRangesToHost = HashMultimap.create();
-        for (InetSocketAddress host : getCachedServers()) {
-            try (CassandraClient client = CassandraClientFactory.getClientInternal(host, config)) {
+        Multimap<Set<TokenRange>, CassandraServer> tokenRangesToHost = HashMultimap.create();
+        for (CassandraServer host : getCachedServers()) {
+            try (CassandraClient client = CassandraClientFactory.getClientInternal(host.proxy(), config)) {
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:511]: `CassandraClientFactory` basically has not changed at all. Maybe sneaky in the way that we are hard binding a proxy Ip to a cass server ip but we will be changing this algorithm anyway. We can also change `CassandraClientFactory` to look uniform later.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 5]
```diff
@@ -264,50 +294,75 @@ private InetSocketAddress getAddressForHostThrowUnchecked(String host) {
...
-        for (InetSocketAddress address : allKnownHosts) {
-            if (Objects.equals(address.getAddress(), resolvedHost)) {
-                return address;
+        for (CassandraServer server : knownNodes) {
+            if (Objects.equals(server.cassandraHostAddress().getAddress(), cassHostAddress.getAddress())) {
+                return server;
             }
         }
 
+        return getCassandraServer(cassHostAddress);
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:314]: hmm we are only comparing the cass hosts above - so reaching this point would mean that we saw a new cass host in the describe_ring op.
okay i will move that super noisy log line here and log conditionally if the proxies do not match.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 6]
```diff
@@ -181,21 +181,51 @@ public void close() {}
...
+                .cassandraHostAddress(cassandraHost)
+                .reachableProxyIps(getReachableProxies(cassandraHost))
+                .build();
+    }
+
+    private List<InetSocketAddress> getReachableProxies(InetSocketAddress addr) {
+        try {
+            // todo(snanda): Remove the log if algorithm changes
+            List<InetSocketAddress> reachableProxies = getReachableProxies(addr.getHostString());
+            log.info(
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:210]: only added for RC - will resolve this comment as soon as i cut an RC

**Review by jeremyk-91 — APPROVED** [review]
Bunch of small things, but I'm fairly convinced that this works

**Comment by jeremyk-91** [comment]
note:
- up to this point, we populate the full list of proxies in `CassandraServer`, but only expose `proxy[0]` to the rest of the underlying code
- this is assumed to be stable
- and the check for absent etc. is based on `CassandraServer` (with some maybe noisy logging for when the server is the same one but the proxies don't match, but we're looking to validate that on a development stack).

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/HostPartitioner.java [hunk 7]
```diff
@@ -58,12 +58,13 @@ private HostPartitioner() {
...
             partitionedByKey.put(ByteBuffer.wrap(keyExtractor.apply(value)), value);
         }
-        ListMultimap<InetSocketAddress, V> valuesByHost = ArrayListMultimap.create();
+        ListMultimap<CassandraServer, V> valuesByHost = ArrayListMultimap.create();
         for (ByteBuffer key : partitionedByKey.keySet()) {
             Preconditions.checkState(key.hasArray(), "Expected an array backed buffer");
             Preconditions.checkState(key.arrayOffset() == 0, "Buffer array must have no offset");
             Preconditions.checkState(key.limit() == key.array().length, "Array length must match the limit");
-            InetSocketAddress host = clientPool.getRandomHostForKey(key.array());
+            // todo(snanda): make this uniform
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/HostPartitioner.java]: Yeah basically I had to change the apis to use `CassServer`, the todo was not required

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java [hunk 8]
```diff
@@ -92,34 +93,40 @@ private boolean isHostHealthy(CassandraClientPoolingContainer container) {
...
     }
 
-    public void remove(InetSocketAddress host) {
+    public void remove(CassandraServer host) {
         blacklist.remove(host);
-        log.info("Remove blacklisted host '{}'", SafeArg.of("removedHost", CassandraLogHelper.host(host)));
+        log.info(
+                "Remove blacklisted host '{}'",
+                SafeArg.of("badHost", CassandraLogHelper.cassandraServer(host)),
+                SafeArg.of("proxy", CassandraLogHelper.host(host.proxy())));
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java:129]: great catch, esp in this PR

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 9]
```diff
@@ -264,50 +294,75 @@ private InetSocketAddress getAddressForHostThrowUnchecked(String host) {
...
-            return new InetSocketAddress(resolvedHost, Iterables.getOnlyElement(allKnownPorts));
+            return Iterables.getOnlyElement(allKnownPorts);
         } else {
-            throw new UnknownHostException("Couldn't find the provided host in server list or current servers");
+            throw new UnknownHostException("No single known port");
         }
     }
 
-    private List<InetSocketAddress> getHostsFor(byte[] key) {
+    private Set<InetSocketAddress> getAllKnownHosts() {
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:315]: The tests were failing over something else. Why does this one look sus?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:397]: 🔥 ok then!

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java [hunk 10]
```diff
@@ -462,11 +473,11 @@ private void verifyNumberOfAttemptsOnHost(
                         Mockito.<FunctionCheckedException<CassandraClient, Object, RuntimeException>>any());
     }
 
-    private CassandraClientPoolImpl clientPoolWithServers(Set<CassandraServer> servers) {
+    private CassandraClientPoolImpl clientPoolWithServers(Set<InetSocketAddress> servers) {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java:476]: nit: clientPoolWithProxies maybe?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:315]: It was just non-trivial to trace where these methods were coming from, and the logic in this return statement in particular is quite magical.

The `currentPools.keySet().stream().map(CassandraServer::proxy).collect(Collectors.toSet())` part could possibly be extracted into `getProxiesFromCurrentPools()`.

Also, isn't `config.servers().accept(new ThriftHostExtractingVisitor()` already a method somewhere? Or has that gone now?

With that, we'd just have `return ImmutableSet.copyOf(Sets.union(getProxiesFromCurrentPools(), getHostsFromConfig())`, which is easier to understand.

**Review by gsheasby — APPROVED** [review]
LGTM, obviously pending:
1 - testing (!)
2 - removing or limiting the potentially noisy log lines

**Review by Jolyon-S — CHANGES_REQUESTED** [review]
I think there are a couple of places where we're recomputing the same DNS lookups for a given IP address. I don't think this affects correctness, but we can probably just do better. Still reviewing the rest.

[EDIT] Or maybe not, but it's likely that we can refactor the code to make this clearer!

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 11]
```diff
@@ -431,9 +431,7 @@ void runOneTimeStartupChecks() {
 
         while (true) {
             if (log.isTraceEnabled()) {
-                log.trace(
-                        "Running function on host {}.",
-                        SafeArg.of("host", CassandraLogHelper.cassandraServer(req.getCassandraServer())));
+                log.trace("Running function on host {}.", SafeArg.of("server", req.getCassandraServer()));
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:434]: nit: is it a host or a server?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 12]
```diff
@@ -550,6 +548,11 @@ private void sanityCheckRingConsistency() {
                 "Cassandra does not appear to have a consistent ring across all of its nodes. This could cause us to"
                         + " lose writes. The mapping of token ranges to hosts is:\n{}",
                 UnsafeArg.of("tokenRangesToHost", CassandraLogHelper.tokenRangesToServer(tokenRangesToHost)),
+                SafeArg.of(
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:551]: this seems like it'll be very big - will that cause memory issues?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java [hunk 13]
```diff
@@ -18,30 +18,27 @@
...
 public interface CassandraServer {
     @Value.Parameter
-    InetSocketAddress cassandraHostAddress();
+    String cassandraHostName();
 
     /**
-     * We do maintain a list of all IPs but do not create a client pool for each one of these. As of now this list is
-     * unused.
+     * {@code cassandraHostName()} with {@code reachableProxyIps()} form one reachable Cassandra server.
+     * While we maintain set of all IPs but do not create a client pool for each one of these.
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:31]: this needs rewording.
```suggestion
     * We maintain set of all IPs for this server, but do not create a client pool for each one.
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java [hunk 14]
```diff
@@ -18,30 +18,27 @@
...
      * The only proxy that will be used to reach the Cassandra host.
-     *
-     * We are making the assumption here that the list of IPs for a host will be consistent.
-     * In case this does not happen, we will black list this host as it will not be reachable using the proxy.
      * */
     @Value.Lazy
     default InetSocketAddress proxy() {
-        return reachableProxyIps().get(0);
+        // we know the set of proxies contains at least one element
+        return reachableProxyIps().stream().findFirst().get();
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:41]:
```suggestion
        return reachableProxyIps().stream().findFirst().orElseThrow();
```

Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:41]: Also, how about adding a `@Value.Check` to disallow setting `reachableProxyIps` to the empty set?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 15]
```diff
@@ -190,23 +190,25 @@ public void close() {}
         }
     }
 
+    /**
+     * It is expected that config provides list of servers that are directly reachable and do not require special IP
+     * resolution.
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:195]: What happens if config does not? Do we just crash and burn?

Can we detect this case, and if so, can we log something useful?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 16]
```diff
@@ -276,28 +278,22 @@ private CassandraServer getAddressForHostThrowUnchecked(String host) {
...
+        Map<String, String> hostnamesByIp = hostnameByIpSupplier.get();
+        String cassandraHostName = hostnamesByIp.getOrDefault(inputHost, inputHost);
+        return CassandraServer.builder()
+                .cassandraHostName(cassandraHostName)
+                .addAllReachableProxyIps(getReachableProxiesThrowUnchecked(cassandraHostName))
+                .build();
+    }
 
-        return getCassandraServer(cassHostAddress);
+    private List<InetSocketAddress> getReachableProxiesThrowUnchecked(String inputHost) throws UnknownHostException {
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:289]: what does the "throw unchecked" mean here?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java [hunk 17]
```diff
@@ -160,7 +160,7 @@ public void shouldRetryOnSameNodeToFailureAndThenRedirect() {
         int numHosts = 4;
         List<CassandraServer> serverList = new ArrayList<>();
         for (int i = 0; i < numHosts; i++) {
-            serverList.add(CassandraServer.from(new InetSocketAddress(i)));
+            serverList.add(CassandraServer.from(InetSocketAddress.createUnresolved(Integer.toString(i), i)));
```
Comment (gsheasby) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java:163]: nit: perhaps extract a method here that you can then use in WeightedServersTest

**Review by gsheasby — COMMENTED** [review]
Generally looks good, I have a few nits, and there's one piece in particular that needs to be reworded.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java (hunk 12 above)
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:551]: _Maybe_ but at this point we are in an inconsistent state and should not be writing to Cass anyway

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java (hunk 14 above)
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:41]: we have that :)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java (hunk 15 above)
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:195]: well we will not be able to talk to Cassandra at all, think those are logged as somewhat meaningful messages already

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java (hunk 16 above)
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:289]: err i killed the other method so this makes no sense now, will rename

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java [hunk 18]
```diff
@@ -214,6 +214,10 @@ private void runTwoNoopsOnTwoHostsAndThrowFromSecondRunOnFirstHost(Exception exc
                 .isInstanceOf(Exception.class);
     }
 
+    private CassandraServer getServerForIndex(int index) {
+        return CassandraServer.from(InetSocketAddress.createUnresolved(Integer.toString(index), index));
```
Comment (gsheasby) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java:218]: I meant to make this (public/package) static, and then use in other places where you're doing the `"0", 0` thing - but certainly not blocking on this! Fine to test/merge as-is.

**Review by gsheasby — APPROVED** [review]

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Set;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface CassandraServer {
+    @Value.Parameter
+    String cassandraHostName();
+
+    /**
+     * {@code cassandraHostName()} with {@code reachableProxyIps()} form one reachable Cassandra server.
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:30]: form :)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 20]
```diff
@@ -264,50 +277,61 @@ private InetSocketAddress getAddressForHostThrowUnchecked(String host) {
     }
 
     @VisibleForTesting
-    InetSocketAddress getAddressForHost(String inputHost) throws UnknownHostException {
-        if (config.addressTranslation().containsKey(inputHost)) {
-            return config.addressTranslation().get(inputHost);
-        }
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:270]: Does this make sense now that we are moving to Rubix?

[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:270]: plus we have the persisted hostIp -> hostname table

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraAbsentHostTracker.java [hunk 21]
```diff
@@ -33,60 +33,62 @@
     private static final SafeLogger log = SafeLoggerFactory.get(CassandraAbsentHostTracker.class);
 
     private final int absenceLimit;
-    private final Map<InetSocketAddress, PoolAndCount> absentHosts;
+    private final Map<CassandraServer, PoolAndCount> absentCassandraServers;
 
     public CassandraAbsentHostTracker(int absenceLimit) {
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraAbsentHostTracker.java:38]: Still have the situation where pod in k8s changes IP upon bounce - I think it is useful to have the absent host tracker

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 22]
```diff
@@ -283,25 +282,24 @@ static int getMaxTriesTotal() {
...
+    Set<CassandraServer> getLocalHosts() {
         return cassandra.getLocalHosts();
     }
 
     private synchronized void refreshPool() {
         blacklist.checkAndUpdate(cassandra.getPools());
 
-        Set<InetSocketAddress> resolvedConfigAddresses =
-                config.servers().accept(new CassandraServersConfigs.ThriftHostsExtractingVisitor());
+        Set<CassandraServer> resolvedConfigAddresses = cassandra.getInitialServerList();
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:302]: inlined

**Comment by jeremyk-91** [comment]
pending smoketest, fine to go in current state 👍

**Pull request merged**